native/dbdsqr.go

```diff
@@ -0,0 +1,493 @@
+	const (
+		hndrth    = 1.0 / 100
+		negEighth = -1.0 / 8
+		maxIter   = 6
+	)
+	if n == 0 {
+		return true
+	}
+	rotate := ncvt > 0 || nru > 0 || ncc > 0
+	if n == 1 {
+		goto OneSixty
+	}
+	// Create a block to help with the above goto and initializaiton
+	{
+		// If the singular vectors do not need to be computed, use qd algorithm.
+		if !rotate {
+			info = impl.Dlasq1(n, d, e, work)
+			// If info is 2 dqds didn't finish, and so try to.
+			if info != 2 {
+				return info == 0
+			}
+			info = 0
+		}
+		nm1 := n - 1
+		nm12 := nm1 + nm1
+		nm13 := nm12 + nm1
+		idir := 0
+
+		eps := dlamchE
+		unfl := dlamchS
+		lower := uplo == blas.Lower
+		var cs, sn, r float64
+		if lower {
+			for i := 0; i < n-1; i++ {
+				cs, sn, r = impl.Dlartg(d[i], e[i])
+				d[i] = r
+				e[i] = sn * d[i+1]
+				d[i+1] *= cs
+				work[i] = cs
+				work[nm1+i] = sn
+			}
+			if nru > 0 {
+				impl.Dlasr(blas.Right, lapack.Variable, lapack.Forward, nru, n, work[0:], work[n-1:], u, ldu)
+			}
+			if ncc > 0 {
+				impl.Dlasr(blas.Left, lapack.Variable, lapack.Forward, n, ncc, work[0:], work[n-1:], c, ldc)
+			}
+		}
+		// Compute singular values to a relative accuracy of tol. If tol is negative
+		// the values will be computed to an absolute accuracy of math.Abs(tol) * norm(b)
+		tolmul := math.Max(10, math.Min(100, eps*negEighth))
+		tol := tolmul * eps
+		var thresh float64
+		var smax float64
+		for i := 0; i < n; i++ {
+			smax = math.Max(smax, math.Abs(d[i]))
+		}
+		for i := 0; i < n-1; i++ {
+			smax = math.Max(smax, math.Abs(e[i]))
+		}
+		var smin1 float64
+		if tol >= 0 {
+			sminoa := math.Abs(d[0])
+			if sminoa != 0 {
+				mu := sminoa
+				for i := 1; i < n; i++ {
+					mu = math.Abs(d[i]) * (mu / (mu + math.Abs(e[i-1])))
+					sminoa = math.Min(sminoa, mu)
+					if sminoa == 0 {
+						break
+					}
+				}
+			}
+			sminoa = sminoa / math.Sqrt(float64(n))
+			thresh = math.Max(tol*sminoa, float64(maxIter*n*n)*unfl)
+		} else {
+			thresh = math.Max(math.Abs(tol)*smax, float64(maxIter*n*n)*unfl)
+		}
+		// Prepare for the main iteration loop for the singular values.
+		maxIt := maxIter * n * n
+		iter := 0
+		oldll := -1
+		oldm := -1
+		// m points to the last element of unconverged part of matrix.
+		m := n
+
+	Outer:
+		for {
+			if m <= 1 {
+				break
+			}
+			if iter > maxIt {
+				info = 0
+				for i := 0; i < n-1; i++ {
+					if e[i] != 0 {
+						info++
+					}
+				}
+				return info == 0
+			}
+			// Find diagonal block of matrix to work on.
+			if tol < 0 && d[m-1] <= thresh {
+				d[m-1] = 0
+			}
+			smax = math.Abs(d[m-1])
+			smin := smax
+			var ll int
+			var broke bool
+			for lll := 0; lll < m-1; lll++ {
+				ll = m - lll - 2
+				abss := math.Abs(d[ll])
+				abse := math.Abs(e[ll])
+				if tol < 0 && abss <= thresh {
+					d[ll] = 0
+				}
+				if abse <= thresh {
+					broke = true
+					break
+				}
+				smin = math.Min(smin, abss)
+				smax = math.Max(math.Max(smax, abss), abse)
+			}
+			if broke {
```

Comment:
Then l2++ goes into the executed break condition and there is no other code. Even if the simplification I'm suggesting doesn't happen that should - at the moment it does nothin to l2 if broke, otherwise it decreases l2, then increases l2. It would be cleaner to not have the else block and put the l2++ in the if block.

But once you've done that you see that the transformation I'm suggesting is obvious.